### PR TITLE
JSDK-2450 track switchoff on higher objects

### PR DESCRIPTION
**Pull request by makarandp0**
bandwidth optimization can cause some track to get switched off/on to manage the available bandwidth. 

We [introduced](https://github.com/twilio/twilio-video.js/pull/682) trackSwitchOn/Off events to notify the developer of such changes. This change propagates those events to higher level objects to be consistent with other events and it makes it bit easy to consume those events.

**Contributing to Twilio**

> All third-party contributors acknowledge that any contributions they provide will be made under the same open-source license that the open-source project is provided under.

TODO:
- [ ] Add changelog entry

- [xI acknowledge that all my contributions will be made under the project's license.

### CODE REVIEW

File: lib/media/track/remotetrackpublication.js [hunk 1]
```diff
@@ -142,6 +146,16 @@ class RemoteTrackPublication extends TrackPublication {
  * @event RemoteTrackPublication#trackEnabled
  */
 
+/**
+ * The {@link RemoteTrack} was switched off.
+ * @event RemoteTrackPublication#trackSwitchedOff
```
Comment (manjeshbhargav) [inline lib/media/track/remotetrackpublication.js]: @makarandp0 The `track` should be an argument to the event handler.

File: lib/media/track/remotetrackpublication.js [hunk 2]
```diff
@@ -142,6 +146,16 @@ class RemoteTrackPublication extends TrackPublication {
...
+ */
+
+/**
+ * The {@link RemoteTrack} was switched on.
+ * @event RemoteTrackPublication#trackSwitchedOn
```
Comment (manjeshbhargav) [inline lib/media/track/remotetrackpublication.js]: @makarandp0 The `track` should be an argument to the event handler.

File: lib/remoteparticipant.js [hunk 3]
```diff
@@ -202,6 +206,22 @@ class RemoteParticipant extends Participant {
  * @event RemoteParticipant#trackSubscriptionFailed
  */
 
+/**
+ * A {@link RemoteParticipant}'s {@link RemoteTrack} was subscribed to.
```
Comment (manjeshbhargav) [inline lib/remoteparticipant.js]: @makarandp0 `s/subscribed to/switched off`.

File: lib/remoteparticipant.js [hunk 4]
```diff
@@ -202,6 +206,22 @@ class RemoteParticipant extends Participant {
...
+ * @param {RemoteTrackPublication} publication - The {@link RemoteTrackPublication}
+ *   for the {@link RemoteTrack} that was switched off
+ * @event RemoteParticipant#trackSwitchedOff
+ */
+
+/**
+ * A {@link RemoteParticipant}'s {@link RemoteTrack} was switched on.
+ * @param {RemoteTrack} track - The {@link RemoteTrack} that was subscribed to
+ * @param {RemoteTrackPublication} publication - The {@link RemoteTrackPublication}
+ *   for the {@link RemoteTrack} that was switched on.
```
Comment (manjeshbhargav) [inline lib/remoteparticipant.js]: @makarandp0 Nitpick: Remove `.`

File: test/integration/spec/connect.js [hunk 5]
```diff
@@ -1271,6 +1271,59 @@ describe('connect', function() {
         });
       });
     });
+
+    if (defaults.topology !== 'peer-to-peer') {
```
Comment (manjeshbhargav) [inline test/integration/spec/connect.js]: @makarandp0 This test should be in `test/integration/spec/room.js`. Also, can you test whether the event is fired on `RemoteTrackPublication` and `RemoteParticipant` as well?

File: test/integration/spec/connect.js [hunk 6]
```diff
@@ -1271,6 +1271,59 @@ describe('connect', function() {
...
+      describe('trackSwitchOff event', () => {
```
Comment (manjeshbhargav) [inline test/integration/spec/connect.js]: @makarandp0 "trackSwitch**ed**Off".

File: test/integration/spec/connect.js [hunk 7]
```diff
@@ -1271,6 +1271,59 @@ describe('connect', function() {
...
+        it('fires on track, trackSubscription, participant and Room object', async () => {
```
Comment (manjeshbhargav) [inline test/integration/spec/connect.js]: @makarandp0 "trackSwitchedOn" test?

File: lib/media/track/remotetrackpublication.js [hunk 8]
```diff
@@ -87,6 +90,7 @@ class RemoteTrackPublication extends TrackPublication {
         isSwitchedOff = signaling.isSwitchedOff;
         if (this.track) {
           this.track._setSwitchedOff(signaling.isSwitchedOff);
+          this.emit(signaling.isSwitchedOff ? 'trackSwitchedOff' : 'trackSwitchedOn');
```
Comment (manjeshbhargav) [inline lib/media/track/remotetrackpublication.js]: @makarandp0 `this.emit(..., track);`.

File: test/integration/spec/connect.js (hunk 5 above)
Comment (makarandp0) [reply test/integration/spec/connect.js]: I am already verifying that event fires on track/RemoteTrackPublication/RemoteParticipant and room. But sure, I will move it to room.js,

**Comment by makarandp0** [comment]
They @manjeshbhargav I have incorporated your feedback. Can you take another look?

File: test/integration/spec/room.js [hunk 9]
```diff
@@ -429,4 +432,113 @@ describe('Room', function() {
...
+      remoteCharlie = aliceRoom.participants.get(charlie.sid);
+    });
+
+    it('trackSwitchedOff fires on track, trackSubscription, participant and Room object', async () => {
+      // listen for switch off event on bob's track on various objects
+      // 1) Track
+      const p1 = new Promise(resolve => loPriTrackPub.track.once('switchedOff', resolve));
+
+      // 2) TrackPublication
+      const p2 = new Promise(resolve => loPriTrackPub.once('trackSwitchedOff', (track) => {
```
Comment (manjeshbhargav) [inline test/integration/spec/room.js]: @makarandp0 Nitpick `track =>`?

File: test/integration/spec/room.js [hunk 10]
```diff
@@ -429,4 +432,113 @@ describe('Room', function() {
...
+
+      // wait for Bob's track getting switched off.
+      await trackSwitchoffPromise;
+
+      // Now listen for switch on event on bob's track on various objects
+      // 1) Track
+      const p1 = new Promise(resolve => loPriTrackPub.track.once('switchedOn', resolve));
+
+      // 2) TrackPublication
+      const p2 = new Promise(resolve => loPriTrackPub.once('trackSwitchedOn', (track) => {
```
Comment (manjeshbhargav) [inline test/integration/spec/room.js]: @makarandp0 Nitpick `track =>`?

**Pull request merged**